trinity/plugins/builtin/json_rpc/plugin.py

```diff
@@ -0,0 +1,122 @@
+        arg_parser.add_argument(
+            "--disable-rpc",
+            action="store_true",
+            help="Disables the JSON-RPC Server",
+        )
+
+    def start(self) -> None:
+        self.logger.info('JSON-RPC Server started')
+        loop = asyncio.get_event_loop()
+        self.context.event_bus.connect()
+
+        db_manager = create_db_manager(self.context.chain_config.database_ipc_path)
+        db_manager.connect()
+
+        chain_class = self.context.chain_config.node_class.chain_class
+        db = db_manager.get_db()  # type: ignore
+        chain = chain_class(db)
+
+        rpc = RPCServer(chain, self.context.event_bus)
+        ipc = IPCServer(rpc, self.context.chain_config.jsonrpc_ipc_path)
+
+        asyncio.ensure_future(ipc.run())
+        loop.run_forever()
+        loop.close()
+
+
+class JsonRpcServerLightPlugin(BasePlugin):
+    """
+    The ``JsonRpcServerLightPlugin`` is an intermediate step to keep the
+    JSON-RPC server inside the networking process when in ``light`` mode.
+    This is because in ``light`` mode, the chain is more coupled to the
+    ``PeerPool``. This should move when the ``PeerPool`` becomes more
+    multiprocess friendly, exposing events and APIs via event bus.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.peer_pool: PeerPool = None
+        self.cancel_token: CancelToken = None
+        self.chain: BaseChain = None
+
+    @property
+    def name(self) -> str:
+        return "JSON-RPC Server Light"
+
+    def should_start(self) -> bool:
+        return all((self.peer_pool is not None,
+                    self.chain is not None,
+                    not self.context.args.disable_rpc,
+                    self.context.args.sync_mode == SYNC_LIGHT))
+
+    def handle_event(self, activation_event: BaseEvent) -> None:
+        if isinstance(activation_event, ResourceAvailableEvent):
+            if activation_event.resource_type is PeerPool:
+                self.peer_pool, self.cancel_token = activation_event.resource
+            elif activation_event.resource_type is BaseChain:
+                self.chain = activation_event.resource
+
+    def configure_parser(self, arg_parser: ArgumentParser, subparser: _SubParsersAction) -> None:
+        # The JsonRpcServerPlugin plugin above already configures the --disable-rpc flag
+        pass
+
+    def start(self) -> None:
+        self.logger.info('JSON-RPC Server started')
+
+        rpc = RPCServer(self.chain, self.context.event_bus)
+        # The IPCServer used to have its own event loop because a comment indicated
+        # it may otherwise run into deadlocks. Not sure yet what to do about that
+        # because having two distinct event loops seems to be problematic now that
+        # the RPCServer uses the eventbus to retrieve the peer count
+        ipc = IPCServer(rpc, self.context.chain_config.jsonrpc_ipc_path)
+        asyncio.ensure_future(ipc.run())
```

Comment:
This is the part that I'm most unsure about. This is in the `light` plugin - the one that remains in the `networking` process. Previously, the IPC server had its own event loop running in another thread. One way we could use to make that work is to have another endpoint (shared plugins only have one single `networking` endpoint so far) end then call `another_endpoint.connect(ipc_loop))`. I tried that end it works in theory but I wasn't sure if this is a real issue to invest time in.

I'd rather invest time into making sure we can get rid of this extra plugin sooner than later. Not sure.

Reply:
We can do dirty intermediate stuff to keep this working but I'm still in the camp that it needs to remain functional.

Reply:
I'm totally on board. It is functional and I'm not sure how I would prove that the extra thread is needed. Seems like something that may only become obvious in longer real world testing?